File: backend/app/api/time_entries.py
from datetime import datetime

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.dependencies import get_async_session, get_current_user
from app.models import User
from app.schemas.time_entry import TimeEntryCreate, TimeEntryRead, TimeEntryUpdate
from app.services.time_entry_service import TimeEntryService

router = APIRouter()


@router.get("/", response_model=list[TimeEntryRead])
async def list_time_entries(
    project_ids: list[int] | None = Query(default=None),
    date_from: datetime | None = Query(default=None),
    date_to: datetime | None = Query(default=None),
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session),
) -> list[TimeEntryRead]:
    service = TimeEntryService(session)
    entries = await service.list_for_user(
        current_user,
        project_ids=project_ids,
        date_from=date_from,
        date_to=date_to,
    )
    return [TimeEntryRead.model_validate(entry) for entry in entries]


@router.post("/", response_model=TimeEntryRead, status_code=status.HTTP_201_CREATED)
async def create_time_entry(
    payload: TimeEntryCreate,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session),
) -> TimeEntryRead:
    service = TimeEntryService(session)
    entry = await service.create(current_user, payload)
    return TimeEntryRead.model_validate(entry)


@router.patch("/{entry_id}", response_model=TimeEntryRead)
async def update_time_entry(
    entry_id: int,
    payload: TimeEntryUpdate,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session),
) -> TimeEntryRead:
    service = TimeEntryService(session)
    entry = await service.update(current_user, entry_id, payload)
    return TimeEntryRead.model_validate(entry)


@router.delete("/{entry_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_time_entry(
    entry_id: int,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_async_session),
) -> None:
    service = TimeEntryService(session)
    await service.delete(current_user, entry_id)

File: frontend/src/services/timeEntries.ts
import { apiClient } from './api-client'
import type { TimeEntry, TimeEntryPayload } from '@/types'

export interface TimeEntryFilters {
  project_ids?: number[]
  date_from?: string
  date_to?: string
}

export const timeEntriesApi = {
  list: async (filters: TimeEntryFilters = {}): Promise<TimeEntry[]> => {
    const params = new URLSearchParams()
    if (filters.project_ids) {
      filters.project_ids.forEach((id) => params.append('project_ids', String(id)))
    }
    if (filters.date_from) params.set('date_from', filters.date_from)
    if (filters.date_to) params.set('date_to', filters.date_to)
    const query = params.toString()
    const url = query ? `/time-entries?${query}` : '/time-entries'
    const { data } = await apiClient.get<TimeEntry[]>(url)
    return data
  },
  create: async (payload: TimeEntryPayload): Promise<TimeEntry> => {
    const { data } = await apiClient.post<TimeEntry>('/time-entries', payload)
    return data
  },
  update: async (
    entryId: number,
    payload: Partial<TimeEntryPayload>,
  ): Promise<TimeEntry> => {
    const { data } = await apiClient.patch<TimeEntry>(
      `/time-entries/${entryId}`,
      payload,
    )
    return data
  },
  remove: async (entryId: number): Promise<void> => {
    await apiClient.delete(`/time-entries/${entryId}`)
  },
}
